### ISSUE TITLE
Code-fix for RS0017

### ISSUE THREAD

**Issue by tobias-tengler** [dotnet/roslyn-analyzers#5130]
### Analyzer

**Diagnostic ID**: [RS0017](https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/Microsoft.CodeAnalysis.PublicApiAnalyzers.md#rs0017-remove-deleted-types-and-members-from-the-declared-api)

### Describe the improvement

Provide a code-fix for the removal of APIs (`RS0017`).

### Describe suggestions on how to achieve the rule

The code-fix would
- remove the line from `PublicAPI.Unshipped.txt` (if it was a new unshipped API only defined in `PublicAPI.Unshipped.txt`)
- place a copy of the line prefixed with `*REMOVED*` in the `PublicAPI.Unshipped.txt` file (if it was a shipped API defined in `PublicAPI.Shipped.txt`)

### Additional context

We're in the process of integrating the PublicApiAnalyzer into [Hot Chocolate](https://github.com/ChilliCream/hotchocolate). 
Currently we invoke the `RS0016` code-fix using [dotnet format](https://github.com/dotnet/format):
```bash
dotnet format --fix-analyzers warn --diagnostics RS0016
```
This allows contributors that do not use Visual Studio 2019 (Rider or VS Code) to invoke the code-fix as well.
While working on a new major version, we noticed how tedious it is to remove existing APIs from these files manually. Being able to do this automatically would drastically simplify our workflow with this analyzer, since now we could just run:
```bash
dotnet format --fix-analyzers warn --diagnostics RS0016,RS0017
```
This being automatic would also prevent any issues that could arise when doing these changes manually.

I'm already looking into the necessary changes and would be willing to contribute a PR for this.

**Comment by Youssef1313** [comment]
@mavasani Wouldn't writing a codefix here just work? I know there is no IDE support for additional file codefixes, but what about from command-line?

cc @JoeRobich

**Comment by tobias-tengler** [comment]
@jmarolf could you elaborate on why you tagged this as "Blocked"?

**Comment by Youssef1313** [comment]
@tobias-tengler It's blocked since it depends on a feature request on roslyn side.

**Comment by jmarolf** [comment]
@Youssef1313 is right. We cannot write codefixes against additional files today. Until dotnet/roslyn adds this capability this issue is blocked

**Comment by tobias-tengler** [comment]
Ah, I see. Thanks for the quick update guys!
I also assume that the new [Package Validation](https://devblogs.microsoft.com/dotnet/package-validation/) will make the API analyzer largely redundant!?

**Comment by rolfbjarne** [comment]
> [@Youssef1313](https://github.com/Youssef1313) is right. We cannot write codefixes against additional files today. Until dotnet/roslyn adds this capability this issue is blocked

I don't see that there's any fundamental difference between fixing RS0016 (which works fine with "dotnet format ...") and fixing RS0017?

**Comment by rolfbjarne** [comment]
> > [@Youssef1313](https://github.com/Youssef1313) is right. We cannot write codefixes against additional files today. Until dotnet/roslyn adds this capability this issue is blocked
> 
> I don't see that there's any fundamental difference between fixing RS0016 (which works fine with "dotnet format ...") and fixing RS0017?

Ah I see, the problem is that there's no source location: https://github.com/dotnet/roslyn-analyzers/issues/5058#issuecomment-825462586

**Comment by sliekens** [comment]
Here is a pwsh script I vibe coded to fix RS0017 (provided as-is without warranty)

``` pwsh
#!/usr/bin/env pwsh
<#
.SYNOPSIS
Removes obsolete public API entries from PublicAPI.Shipped.txt based on RS0017 warnings.

.DESCRIPTION
This script runs dotnet build, parses RS0017 warnings from the build output,
and removes the corresponding lines from PublicAPI.Shipped.txt files.

.EXAMPLE
.\Remove-ObsoletePublicApis.ps1
#>

[CmdletBinding()]
param()

$ErrorActionPreference = "Stop"

# Run dotnet build with --no-incremental to get RS0017 warnings
Write-Verbose "Running dotnet build --no-incremental..."
$buildOutput = & dotnet build --no-incremental 2>&1
$buildOutputString = $buildOutput -join "`n"

# Parse RS0017 warnings
# RS0017 format: "Symbol 'X.Y.Z' is part of the declared API, but is either not public or could not be found"
$rs0017Pattern = "warning RS0017:.*?Symbol '([^']+)'"
$apisToRemove = [System.Collections.Generic.HashSet[string]]::new([System.StringComparer]::Ordinal)

foreach ($match in [System.Text.RegularExpressions.Regex]::Matches($buildOutputString, $rs0017Pattern, [System.Text.RegularExpressions.RegexOptions]::Singleline)) {
    $apiName = $match.Groups[1].Value
    [void]$apisToRemove.Add($apiName)
    Write-Verbose "Found RS0017 warning for: $apiName"
}

if ($apisToRemove.Count -eq 0) {
    Write-Host "No RS0017 warnings found. No changes needed."
    exit 0
}

Write-Host "Found $($apisToRemove.Count) RS0017 warning(s). Processing PublicAPI.Shipped.txt files..."

# Find all PublicAPI.Shipped.txt files
$publicApiFiles = Get-ChildItem -Path . -Filter "PublicAPI.Shipped.txt" -Recurse

foreach ($file in $publicApiFiles) {
    Write-Verbose "Processing: $($file.FullName)"
    
    $content = Get-Content -Path $file.FullName -Encoding UTF8
    $originalLineCount = $content.Count
    
    # Filter out lines matching the APIs to remove
    $newContent = $content | Where-Object {
        $line = $_
        $shouldKeep = $true
        
        foreach ($api in $apisToRemove) {
            # Check if the line starts with the API name or is an abstract member of the API
            if ($line -eq $api -or $line.StartsWith("$api.") -or $line.StartsWith("abstract $api.")) {
                $shouldKeep = $false
                Write-Host "  Removing: $line"
                break
            }
        }
        
        return $shouldKeep
    }
    
    $newLineCount = if ($newContent -is [array]) { $newContent.Count } else { if ($newContent) { 1 } else { 0 } }
    $removedCount = $originalLineCount - $newLineCount
    
    if ($removedCount -gt 0) {
        Set-Content -Path $file.FullName -Value $newContent -Encoding UTF8
        Write-Host "  Removed $removedCount line(s) from $($file.Name)"
    }
}

Write-Host "Done!"
```
